components/TextFields/src/MDCTextField.m

```diff
@@ -0,0 +1,439 @@
+          ? CGRectGetWidth([self rightViewRectForBounds:bounds])
+          : CGRectGetWidth([self leftViewRectForBounds:bounds]);
+  CGFloat rightViewWidth =
+      self.mdc_effectiveUserInterfaceLayoutDirection == UIUserInterfaceLayoutDirectionRightToLeft
+          ? CGRectGetWidth([self leftViewRectForBounds:bounds])
+          : CGRectGetWidth([self rightViewRectForBounds:bounds]);
+
+  if (self.leftView.superview) {
+    textRect.origin.x += leftViewWidth;
+    textRect.size.width -= leftViewWidth;
+  }
+
+  if (self.rightView.superview) {
+    textRect.size.width -= rightViewWidth;
+    // If there is a rightView, the clearButton will not be shown.
+  } else {
+    CGFloat clearButtonWidth = CGRectGetWidth(self.clearButton.bounds);
+    clearButtonWidth += 2 * MDCTextInputClearButtonImageBuiltInPadding;
+
+    // Clear buttons are only shown if there is entered text or programatically set text to clear.
+    if (self.text.length > 0) {
+      switch (self.clearButtonMode) {
+        case UITextFieldViewModeAlways:
+        case UITextFieldViewModeUnlessEditing:
+          textRect.size.width -= clearButtonWidth;
+          break;
+        default:
+          break;
+      }
+    }
+  }
+
+  // UITextFields have a centerY based layout. And you can change EITHER the height or the Y. Not
+  // both. Don't know why. So, we have to leave the text rect as big as the bounds and move it to a
+  // Y that works.
+  CGFloat actualY =
+      (CGRectGetHeight(bounds) / 2.f) - MDCRint(MAX(self.font.lineHeight,
+                                                    self.placeholderLabel.font.lineHeight) /
+                                                2.f);  // Text field or placeholder
+  actualY = textContainerInset.top - actualY;
+  textRect.origin.y = actualY;
+
+  if (self.mdc_effectiveUserInterfaceLayoutDirection == UIUserInterfaceLayoutDirectionRightToLeft) {
+    // Now that the text field is laid out as if it were LTR, we can flip it if necessary.
+    textRect = MDCRectFlippedForRTL(textRect, CGRectGetWidth(bounds),
+                                    UIUserInterfaceLayoutDirectionRightToLeft);
+  }
+
+  return textRect;
+}
+
+- (CGRect)editingRectForBounds:(CGRect)bounds {
+  // First the textRect is loaded. Then it's shaved for cursor and/or clear button.
+  CGRect editingRect = [self textRectForBounds:bounds];
+
+  // The textRect comes to us flipped for RTL (if RTL) so we flip it back before adjusting.
+  if (self.mdc_effectiveUserInterfaceLayoutDirection == UIUserInterfaceLayoutDirectionRightToLeft) {
+    editingRect = MDCRectFlippedForRTL(editingRect, CGRectGetWidth(bounds),
+                                       UIUserInterfaceLayoutDirectionRightToLeft);
+  }
+
+  // UITextFields show EITHER the clear button or the rightView. If the rightView has a superview,
+  // then it's being shown and the clear button isn't.
+  if (self.rightView.superview) {
```

Comment:
Don't know. Unfortunately since that's is private API, we can't guarantee
that these views will always be added to self. They could introduce an
intermediate layer.